mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -865,4 +934,114 @@ class Completions(
+          Some(m.selector)
+        // foo ma@tch (no cases)
+        case (m @ Match(
+              _,
+              CaseDef(Literal(Constant(null)), _, _) :: Nil,
+            )) :: _ =>
+          Some(m.selector)
+        case _ => None
+  end MatchExtractor
+
+  object CaseExtractor:
+    def unapply(path: List[Tree]): Option[(Tree, Tree)] =
+      path match
+        // foo match
+        // case None => ()
+        // ca@@
+        case (id @ Ident(name)) :: Block(stats, expr) :: parent :: _
+            if "case"
+              .startsWith(name.toString()) && isLastMatch(
+              stats
+            ) && expr == id =>
+          val selector = stats.last.asInstanceOf[Match].selector
+          Some((selector, parent))
+        // List(Option(1)).collect {
+        //   case Some(value) => ()
+        //   ca@@
+        // }
+        case (ident @ Ident(name)) :: Block(
+              _,
+              expr,
+            ) :: (_: CaseDef) :: (m: Match) :: parent :: _
+            if ident == expr && "case"
+              .startsWith(name.toString()) =>
+          Some((m.selector, parent))
+        // foo match
+        //  ca@@
+        case (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // List(foo).map { ca@@ }
+        case (ident @ Ident(name)) :: Block(stats, expr) :: (appl @ Apply(
+              fun,
+              args,
+            )) :: _
+            if stats.isEmpty && ident == expr && "case".startsWith(
+              name.toString()
+            ) =>
+          Some((EmptyTree, appl))
+
+        case _ => None
+  end CaseExtractor
+
+  object CasePatternExtractor:
+    def unapply(path: List[Tree]) =
+      path match
+        // case Som@@
+        case (ident @ Ident(
+              name
+            )) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case abc @ Som@@
+        case (ident @ Ident(name)) :: Bind(
+              _,
+              _,
+            ) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case abc @ @@
+        case Bind(_, _) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        case _ => None
+
+  end CasePatternExtractor
+
+  object TypedCasePatternExtractor:
+    def unapply(path: List[Tree]) =
+      path match
+        // case _: Som@@ =>
+        case (ident @ Ident(name)) :: Typed(
+              _,
+              _,
+            ) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case _: @@ =>
+        case Typed(_, _) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case ab: @@ =>
+        case Bind(
+              _,
+              Typed(_, _),
+            ) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case ab: Som@@ =>
```

Comment:
This is a weird case, why would we want to support this?  Is it for infix types? Are we able sensible show suggestions in that case?

Reply:
I am not sure what is weird in this case? When doing `match` on `Option` `case ab: Som@@ =>` will suggest `Some[_]` and `case ab: @@ =>` `Some[_]` and `None`

Reply:
ach, makes sense